Add loaded dice

Palladinium opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR

This adds a bag full of loaded dice, purchasable under the "Pointless" category of the Syndicate uplink for 1 TC. I considered making them available at an emagged library vending machine instead, but I'm not sure if that's better or worse than the uplink purchase (suggestions welcome).

Loaded dice, and the bag they come in, look exactly like their regular counterparts. When a loaded die is held in the hand or pocket (and only when the bag is also held in the hand or pocket), an extra verb is available, which shows a UI to set a predefined value to roll (this is similar to how chameleon clothing can be changed only while being worn). After setting, the die will always roll the set value.

Alt-clicking the die while in hand (or using Alt-Z), as well as the "unset" button in the UI, can be used to unset the die, which makes it behave as normal (rolling random sides). Dice start unset when spawned.

## Why / Balance

It's a fun, minor roleplaying gimmick with very few balance repercussions. I had a couple of rounds where dice rolls mattered quite a bit, so I figured antags can get some fun with it. Players can cheat at IC tabletop games, win a wager against the HoS for their shotgun, or maybe convince a DM lawset AI to let them use their own dice. That is, until someone catches them.

## Technical details
I added a new `LoadedDiceComponent` which stores the set roll value (if any). When the server-side `DiceSystem` tries to roll a value, it now raises a new `DiceRollEvent`, which the `LoadedDiceSystem` listens for to override the roll. This event should also make future work aiming to influence dice rolls more modular (for example, lucky/unlucky gear which make you roll higher/lower).

The `SharedLoadedDiceSystem` handles a new UI to change the value the die is set to. The UI code is largely based on the chameleon clothes code, but vastly simpler because of the smaller scope of this feature.

## Media
[loaded-dice.webm](https://github.com/user-attachments/assets/11b7cb70-d0c5-491f-aeda-61c647c89955)


## Requirements
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes
No breaking changes (I think)

**Changelog**

:cl:
- add: Added loaded dice, purchasable by syndicate agents under "pointless".

beck-thompson commented:
Love the idea! My only suggestion would be to make it so like you can only change the weighting if you are holding the bag they spawned in or something so it can't easily be metagamed.

lzk228 commented on Content.Server/Dice/DiceSystem.cs:
```diff
@@ -19,7 +19,14 @@ public override void Roll(EntityUid uid, DiceComponent? die = null)
-        var roll = _random.Next(1, die.Sides + 1);
+        int roll;
+
+        if (TryComp<LoadedDiceComponent>(uid, out var loaded) && loaded.SelectedSide != null) {
+            roll = loaded.SelectedSide.Value;
+        } else {
+            roll = _random.Next(1, die.Sides + 1);
+        };
```
we usually do
```suggestion
        if (TryComp<LoadedDiceComponent>(uid, out var loaded) && loaded.SelectedSide != null) 
        {
            roll = loaded.SelectedSide.Value;
        } 
        else
        {
            roll = _random.Next(1, die.Sides + 1);
        }
```
tho i'm not sure
and is there a need in semicolon at the end?

Palladinium replied:
I'm probably going to refactor this into an event following a discussion in the development channel on Discord, but you're right I just forgot to follow code style for a minute there.

lzk228 commented on Resources/Prototypes/Entities/Objects/Fun/dice.yml:
```diff
@@ -131,3 +131,81 @@
+      enum.LoadedDiceUiKey.Key:
+        type: LoadedDiceBoundUserInterface
+
+- type: entity
+  parent: d20Dice
+  id: d20DiceLoaded
+  suffix: loaded
+  components:
+  - type: LoadedDice
+  - type: UserInterface
+    interfaces:
+      enum.LoadedDiceUiKey.Key:
+        type: LoadedDiceBoundUserInterface
+
```
you can do separate parent with both components and suffix and then write `parent: [d20Dice, BaseDiceLoaded]`
it will reduce yml lines

lzk228 commented on Content.Server/Dice/LoadedDiceSystem.cs:
```diff
@@ -0,0 +1,104 @@
+        // There may not be a user if the change is not done by a player, e.g. OnMapInit
+        if (component.User != null)
+        {
+            var dieName = Identity.Entity(uid, EntityManager);
```
is there a reason for identity?
also i suggest to add admin logs for this

Palladinium replied:
I just needed to get the dice entity's visible name, and that's what I could find was used in similar code. Is there a more appropriate function for item-like entities, where `Identity` is overkill?

lzk228 replied:
just putting uid in loc string below should work

Everturning commented:
can you add some sort of inhand interaction to set the dice to unset so that you can cheat better?
"guys im not cheating look"

0x6273 commented on Content.Shared/Dice/LoadedDiceComponent.cs:
```diff
@@ -0,0 +1,57 @@
+using Robust.Shared.GameStates;
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.Dice;
```
This probably shouldn't be in shared, since clients shouldn't know if a dice is loaded or not without holding it.

beck-thompson replied:
I don't think it matters because it already has the loaded dice user interface on it which is in shared + the protoId already gives away that its loaded

0x6273 replied:
Prototype IDs probably shouldn't be sent to clients in release builds. But you're right about the ui, you would need to tell the client about it only when the dice is held. I don't know of a good way to do that.

Palladinium replied:
Related: I was considering (on @beck-thompson's suggestion) moving the `LoadedDiceSystem` to shared in order to avoid the noticeable delay when right-clicking before the special verb appears, but I think in the long run it makes more sense to keep as much of it in the server side and try to keep information secret from clients. That's obviously not fully possible until at least some sort of refactor of the bound UI system that allows for more "private" UI, but that seems like the better direction to aim towards.

Palladinium commented:
> can you add some sort of inhand interaction to set the dice to unset so that you can cheat better? "guys im not cheating look"

You can do that from the same menu where you set it to a value, using the wide "Unset" button at the top. Or do you mean I should add a different verb that doesn't require opening a UI?

Palladinium commented:
> Love the idea! My only suggestion would be to make it so like you can only change the weighting if you are holding the bag they spawned in or something so it can't easily be metagamed. The only issue with that is it would not be clear how it works when you find it in maints or something...

I agree that it's not great that anyone can just nab the die from the ground and check it without much counterplay, at least with chameleon clothes they need to strip you. I'm not a fan of "binding" the dice to the bag (both thematically and mechanically), though I don't really have a better solution either. I think as it is should be probably fine until (if) we figure out a better system for covert/stealth items.

Everturning commented:
> > can you add some sort of inhand interaction to set the dice to unset so that you can cheat better? "guys im not cheating look"
> 
> You can do that from the same menu where you set it to a value, using the wide "Unset" button at the top. Or do you mean I should add a different verb that doesn't require opening a UI?

Not a separate verb, but something you can do in-world that sets it back to normal so when people doubt your extreme luck you can just alt-click it and boom, you're not as lucky!

KingOnBudget commented:
OH MY GOD THIS IS GREAT! IMAGINE SOMEBODY USING THIS WITH STORYTELLER AI!

WhiteGorila commented:
"if I roll a 20 I'm going to fucking kill myself"

>Explodes

Palladinium commented:
After re-reading the Metashield rule on Stealth items I think I'm going to remove these from the maints loot pool. If someone were to find them in maints, they'd need to break the stealth item rule and metagaming their purpose in order to actually get any use from them, at which point they may as well be regular dice.

IamVelcroboy commented on Resources/Prototypes/Entities/Objects/Fun/dice.yml:
```diff
@@ -131,3 +131,44 @@
+- type: entity
+  abstract: true
+  id: BaseDiceLoaded
+  suffix: loaded
```
Generally, suffixes are upper case for consistency.

VlaDOS1408 commented:
I think it's possible to add them to a hacked vending machine in the library. They will be used more often

inewma0904 commented:
I honestly think it would be better if the dice were visually identical to regular dice. It would VASTLY simplify the metashield problems, and allow them to appear in maints. Alternatively, the menu to alter the dice chances could appear when a screwdriver is used on them, similar to hacking a vending machine.

0x6273 commented:
Maybe it could be voice activated instead of having an UI? So you can whisper a number to it and then it rolls that number, and if you don't say a number it behaves like a normal dice.

inewma0904 commented:
I kind of like the whispering idea, given how it fits with the gambling theme.

Everturning commented:
I think you're making it a bit too complex for a 1tc item

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+
+    private void OnVerb(EntityUid uid, LoadedDiceComponent component, GetVerbsEvent<InteractionVerb> args)
+    {
+        if (!args.CanAccess || !args.CanInteract || component.User != args.User)
```
Two things about this:
1.) This trick I learned! Use `TryGetContainingContainer` from `SharedContainerSystem`! See #30805 for a simple example. This will allow you to remove all this pesky logic of keeping track of who the user is. You might want to make it into a function as well because you use the user a lot but use your best judgment on that

2.) I personally think it might be better if you had to be holding (or have it in your pocket) the dice bag to be able to switch sides. That'll make it a little less easy to metagame! Honestly though, its a 1 TC gimmick item so if you want it to work this way thats fine. The system can easily be changed later

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+        SetSelectedSide(uid, component, null);
+    }
+
+    private void OnVerb(EntityUid uid, LoadedDiceComponent component, GetVerbsEvent<InteractionVerb> args)
```
I'm leaving the comment here but this is for all the event listeners. Usually whats done now is:
`Entity<LoadedDiceComponent> ent, ref GetVerbsEvent<InteractionVerb> args`

Its up to you if you want to change it, there are pros and cons to both (Sloth likes it the way you have because it gets annoying to debug because it always steps in)

0x6273 replied:
It's actually mandatory to use `Entity<TComp>` for new code, according to the announcement: https://discord.com/channels/310555209753690112/928056457117696020/1164654158725525575

> **All of these changes must be present in PRs made and opened after today's engine update. PRs currently open or in progress are not affected.**

Palladinium replied:
Eh, I may as well switch it to the new way

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+
+    private void OnMapInit(EntityUid uid, LoadedDiceComponent component, MapInitEvent args)
+    {
+        SetSelectedSide(uid, component, null);
```
I'm pretty sure this won't do anything?

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+        _uiSystem.SetUiState(uid, LoadedDiceUiKey.Key, state);
+    }
+
+    public void SetSelectedSide(EntityUid uid, LoadedDiceComponent? component = null, int? selectedSide = null)
```
If its a public function (I think its probably a good idea to make it public so keep that!), make sure to add a full comment that includes all the fields etc... Specifically remember to mention that setting selectedSide to null makes it reset

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+        if (!Resolve(uid, ref component, false))
+            return;
+
+        if (!TryComp<DiceComponent>(uid, out var die))
```
You might as well make this an optional parameter as well!

beck-thompson commented on Content.Client/Dice/LoadedDiceSystem.cs:
```diff
@@ -0,0 +1,10 @@
+///     Stub system to ensure client prediction of <see cref="SharedLoadedDiceSystem"/>.
+/// </summary>
+public sealed class LoadedDiceSystem : SharedLoadedDiceSystem
+{
```
I personally would change this to:
`public sealed class LoadedDiceSystem : SharedLoadedDiceSystem;`

Totally up to you though

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+using Content.Shared.Hands;
+using Content.Shared.Popups;
+using Content.Shared.Verbs;
+using Robust.Shared.GameObjects;
```
Unused!

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+        SetSelectedSide(uid, component, args.SelectedSide);
+    }
+
+    private void TryOpenUi(EntityUid uid, EntityUid user, LoadedDiceComponent? component = null)
```
Don't make the `LoadedDiceComponent` nullable and just remove the `Resolve`!

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+        _uiSystem.TryToggleUi(uid, LoadedDiceUiKey.Key, actor.PlayerSession);
+    }
+
+    private void UpdateUi(EntityUid uid, DiceComponent? die = null, LoadedDiceComponent? component = null)
```
Same, this is private so just make the fields non nullable and remove the Resolves

beck-thompson commented on Content.Shared/Dice/SharedLoadedDiceSystem.cs:
```diff
@@ -0,0 +1,141 @@
+                adminMessage = $"{ToPrettyString(component.User):player} set the loaded {ToPrettyString(uid):target} to roll {sideValue}";
+            }
+
+            _popup.PopupEntity(message, uid, component.User.Value, PopupType.Small);
```
Use popup predicted, its running too many times!

![image](https://github.com/user-attachments/assets/b57b02e6-cce7-47e1-9984-67cb8aae2c0c)

beck-thompson reviewed (changes requested):
Looks good just a few minor style / cleanup things! Also, if you drop the dice, but keep the UI open you can change the value while its on the floor. Probably just make the UI close when you drop the dice and/or make it only change values when its being held by the person clicking.

Also the unset
![image](https://github.com/user-attachments/assets/b61eebc0-70f9-420f-80db-ea5e67f4e76f)

TheShuEd commented:
What is the status of this PR? Do you plan to continue the work?

Palladinium commented:
Life happened and this slipped through my fingers. I'll loop around to it over the next couple of days and address the feedback, try to get the ball rolling on this again.

beck-thompson commented:
> Life happened and this slipped through my fingers. I'll loop around to it over the next couple of days and address the feedback, try to get the ball rolling on this again.

Take your time no pressure :)

Palladinium commented:
I'm pretty sure I've addressed all the feedback, and fixed most or all of the issues with this PR, however, due to the above linked issue I can't reliably test it, since all dice rolls are super buggy right now for me due to my laggy virtualized setup.

I'll change this PR to a draft until the linked issue with the dice randomness is fixed, whether by me or someone else.

TheShuEd commented:
Given that this PR has been in a “Awaiting Changes” state for over 2 months without any activity, I'm closing it as derelict. If you would like to continue, please write and the PR will be reopened.
